src/optimagic/constraints.py

```diff
@@ -0,0 +1,246 @@
+            raise InvalidConstraintError("At least two selectors must be provided.")
+
+        if not all(callable(s) for s in self.selectors):
+            raise InvalidConstraintError("All selectors must be callable.")
+
+
+@dataclass(frozen=True)
+class FlatSDCorrConstraint(Constraint):
+    selector: Callable[[PyTree], PyTree] = identity_selector
+    _: KW_ONLY
+    bounds_distance: float | None = None
+
+    def _to_dict(self) -> dict[str, Any]:
+        return {
+            "type": "sdcorr",
+            "selector": self.selector,
+            **_select_non_none(bounds_distance=self.bounds_distance),
+        }
+
+
+@dataclass(frozen=True)
+class LinearConstraint(Constraint):
+    selector: Callable[[PyTree], ArrayLike | "pd.Series[float]" | float | int] = (
+        identity_selector
+    )
+    _: KW_ONLY
+    weights: ArrayLike | "pd.Series[float]" | float | int | None = None
+    lower_bound: float | int | None = None
+    upper_bound: float | int | None = None
+    value: float | int | None = None
+
+    def _to_dict(self) -> dict[str, Any]:
+        return {
+            "type": "linear",
+            "selector": self.selector,
+            "weights": self.weights,
+            **_select_non_none(
+                lower_bound=self.lower_bound,
+                upper_bound=self.upper_bound,
+                value=self.value,
+            ),
+        }
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+
+        if _all_none(self.lower_bound, self.upper_bound, self.value):
+            raise InvalidConstraintError(
+                "At least one of 'lower_bound', 'upper_bound', or 'value' must be "
+                "non-None."
+            )
+        if self.value is not None and not _all_none(self.lower_bound, self.upper_bound):
+            raise InvalidConstraintError(
+                "'value' cannot be used with 'lower_bound' or 'upper_bound'."
+            )
+
+        if not isinstance(self.weights, np.ndarray | list | pd.Series | float | int):
+            raise InvalidConstraintError(
+                "'weights' must be an array-like, a pandas Series, a float, or an int."
+            )
+
+        if self.lower_bound is not None and not isinstance(
+            self.lower_bound, float | int
+        ):
+            raise InvalidConstraintError("'lower_bound' must be a float or an int.")
+
+        if self.upper_bound is not None and not isinstance(
+            self.upper_bound, float | int
+        ):
+            raise InvalidConstraintError("'upper_bound' must be a float or an int.")
+
+        if self.value is not None and not isinstance(self.value, float | int):
+            raise InvalidConstraintError("'value' must be a float or an int.")
+
+
+@dataclass(frozen=True)
+class NonlinearConstraint(Constraint):
+    selector: Callable[[PyTree], PyTree] = identity_selector
+    _: KW_ONLY
+    func: Callable[[PyTree], ArrayLike | "pd.Series[float]" | float] | None = None
```

Comment:
Why `"pd.Series[float]"` instead of `pd.Series[pd.FloatDtype....]`?

Reply:
![image](https://github.com/user-attachments/assets/d9529442-b0c4-46de-949b-f909be143ea2)

Reply:
Mypy can read the information from `"pd.Series[float]"`.